examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt

```diff
@@ -0,0 +1,354 @@
+      READ_STORAGE_PERMISSION_REQUEST_CODE
+    )
+  }
+
+  inner class OtaProviderCallback : OTAProviderDelegate {
+    private var fileName: String? = null
+    private var version: Long = 0
+    private var versionString: String? = null
+    private var path: String? = null
+
+    private var fileInputStream: FileInputStream? = null
+    private var bufferedInputStream: BufferedInputStream? = null
+
+    fun setOTAFile(version: Long, versionString: String, fileName: String, path: String) {
+      this.version = version
+      this.versionString = versionString
+      this.fileName = fileName
+      this.path = path
+    }
+
+    override fun handleQueryImage(
+      vendorId: Int,
+      productId: Int,
+      softwareVersion: Long,
+      hardwareVersion: Int?,
+      location: String?,
+      requestorCanConsent: Boolean?,
+      metadataForProvider: ByteArray?
+    ): OTAProviderDelegate.QueryImageResponse {
+      Log.d(
+        TAG,
+        "handleQueryImage, $vendorId, $productId, $softwareVersion, $hardwareVersion, $location"
+      )
+      return OTAProviderDelegate.QueryImageResponse(version, versionString, fileName)
+    }
+
+    override fun handleOTAQueryFailure(error: Int) {
+      Log.d(TAG, "handleOTAQueryFailure, $error")
+      showMessage("handleOTAQueryFailure : $error")
+      deviceController.finishOTAProvider()
+    }
+
+    override fun handleApplyUpdateRequest(
+      nodeId: Long,
+      newVersion: Long
+    ): OTAProviderDelegate.ApplyUpdateResponse {
+      Log.d(TAG, "handleApplyUpdateRequest, $nodeId, $newVersion")
+      return OTAProviderDelegate.ApplyUpdateResponse(
+        OTAProviderDelegate.ApplyUpdateActionEnum.Proceed,
+        APPLY_WAITING_TIME
+      )
+    }
+
+    override fun handleNotifyUpdateApplied(nodeId: Long) {
+      Log.d(TAG, "handleNotifyUpdateApplied, $nodeId")
+      showMessage("Finish Firmware Update : $nodeId")
+    }
+
+    override fun handleBDXTransferSessionBegin(
+      nodeId: Long,
+      fileDesignator: String?,
+      offset: Long
+    ) {
+      Log.d(TAG, "handleBDXTransferSessionBegin, $nodeId, $fileDesignator, $offset")
+      if (path == null) {
+        Log.d(TAG, "path is null")
+        return
+      }
+      try {
+        val file = File(path!!)
+        fileInputStream = FileInputStream(file)
+        bufferedInputStream = BufferedInputStream(fileInputStream)
+      } catch (e: IOException) {
+        Log.d(TAG, "exception", e)
+        if (fileInputStream != null) {
+          fileInputStream!!.close()
+          fileInputStream = null
+        }
+        if (bufferedInputStream != null) {
+          bufferedInputStream!!.close()
+          bufferedInputStream = null
+        }
+        return
+      }
+    }
+
+    override fun handleBDXTransferSessionEnd(errorCode: Long, nodeId: Long) {
+      Log.d(TAG, "handleBDXTransferSessionEnd, $errorCode, $nodeId")
+      if (bufferedInputStream != null) {
+        bufferedInputStream!!.close()
```

Comment:
Please try to avoid !! across this file.

Reply:
I modifed